opentelemetry/src/trace/span_context.rs

```diff
@@ -94,89 +95,139 @@ impl fmt::LowerHex for TraceFlags {
     }
 }
 
-/// TraceId is an 16-byte value which uniquely identifies a given trace
-/// The actual `u128` value is wrapped in a tuple struct in order to leverage the newtype pattern
+/// A 16-byte value which identifies a given trace.
+///
+/// The id is valid if it contains at least one non-zero byte.
 #[cfg_attr(feature = "serialize", derive(Deserialize, Serialize))]
-#[derive(Clone, Debug, PartialEq, Eq, Copy, Hash)]
-pub struct TraceId(u128);
+#[derive(Clone, PartialEq, Eq, Copy, Hash)]
+pub struct TraceId(pub(crate) u128);
 
 impl TraceId {
-    /// Construct a new invalid (zero-valued) TraceId
-    pub fn invalid() -> Self {
-        TraceId(0)
+    /// Invalid trace id
+    pub const INVALID: TraceId = TraceId(0);
+
+    /// Create a trace id from its representation as a byte array.
+    pub const fn from_bytes(bytes: [u8; 16]) -> Self {
+        TraceId(u128::from_be_bytes(bytes))
     }
 
-    /// Convert from u128 to TraceId
-    pub fn from_u128(item: u128) -> Self {
-        TraceId(item)
+    /// Return the representation of this trace id as a byte array.
+    pub const fn to_bytes(self) -> [u8; 16] {
+        self.0.to_be_bytes()
     }
 
-    /// Convert from TraceId to u128
-    pub fn to_u128(self) -> u128 {
-        self.0
+    /// Converts a string in base 16 to a trace id.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use opentelemetry::trace::TraceId;
+    ///
+    /// assert!(TraceId::from_hex("42").is_ok());
+    /// assert!(TraceId::from_hex("58406520a006649127e371903a2de979").is_ok());
+    ///
+    /// assert!(TraceId::from_hex("not_hex").is_err());
+    /// ```
+    pub fn from_hex(hex: &str) -> Result<Self, ParseIntError> {
+        u128::from_str_radix(hex, 16).map(TraceId)
     }
+}
 
-    /// Convert from TraceId to Hexadecimal String
-    pub fn to_hex(self) -> String {
-        format!("{:032x}", self.0)
+impl From<[u8; 16]> for TraceId {
+    fn from(bytes: [u8; 16]) -> Self {
+        TraceId::from_bytes(bytes)
     }
+}
 
-    /// Convert from TraceId to Big-Endian byte array
-    pub fn to_byte_array(self) -> [u8; 16] {
-        self.0.to_be_bytes()
+impl From<TraceId> for [u8; 16] {
+    fn from(id: TraceId) -> [u8; 16] {
+        id.to_bytes()
     }
+}
```

Comment:
What's the motivation for having this? I think I'd maybe prefer to require callers to spell `id.to_bytes()` explicitly (because this sort of loses semantic context).

Reply:
sounds reasonable 👍